File: javascript/src/credits/types/ledger.ts
import type { Decimal } from "decimal.js";

// ── Canonical ledger history ─────────────────────────────────────────

/** One immutable monetary entry on a canonical credit account. */
export interface LedgerEntry {
  entryId: string;
  accountId: string;
  actorUserId: string | null;
  amount: Decimal;
  entryType: string;
  operation: string;
  referenceEntryId: string | null;
  idempotencyKey: string | null;
  metadata: Record<string, unknown> | null;
  createdAt: string;
}

/** Stable position in an account ledger. */
export interface LedgerCursor {
  createdAt: string;
  entryId: string;
}

/** Cursor-only options for listing account ledger entries. */
export interface ListLedgerEntriesOptions {
  entryTypes?: string[];
  fromDate?: Date;
  toDate?: Date;
  limit?: number;
  cursor?: LedgerCursor | null;
}

/** One stable cursor page. */
export interface LedgerPage {
  items: LedgerEntry[];
  nextCursor: LedgerCursor | null;
}

/**
 * A metered usage charge from the usage charge journal.
 *
 * Usage charges are deliberately separate from ledger entries: included
 * allowance consumption does not create a monetary ledger debit, but it is
 * still a billable usage event that belongs in the usage history.
 */
export interface UsageCharge {
  usageId: string;
  accountId: string;
  operation: string;
  requested: Decimal;
  charged: Decimal;
  allowanceRequested: Decimal;
  allowanceCovered: Decimal;
  billingDisposition: "billable" | "record_only";
  feature: string | null;
  model: string | null;
  region: string | null;
  eventAt: string;
  idempotencyKey: string;
  metadata: Record<string, unknown> | null;
  createdAt: string;
}

export interface UsageChargeCursor {
  eventAt: string;
  usageId: string;
}

export interface ListUsageChargesOptions {
  fromDate?: Date;
  toDate?: Date;
  limit?: number;
  cursor?: UsageChargeCursor | null;
  includeRecordOnly?: boolean;
}

export interface UsageChargePage {
  items: UsageCharge[];
  nextCursor: UsageChargeCursor | null;
}

/** Result of appending a usage event without creating another balance debit. */
interface UsageRecordResultBase {
  userId: string;
  requested: Decimal;
  idempotent: boolean;
}

export interface UsageRecordSuccess extends UsageRecordResultBase {
  error: null;
  usageId: string;
}

export interface UsageRecordFailure extends UsageRecordResultBase {
  error: string;
  usageId: null;
  idempotent: false;
}

export type UsageRecordResult = UsageRecordSuccess | UsageRecordFailure;

/** Cursor-only options for the usage-only ledger view. */
export interface ListUsageEntriesOptions {
  fromDate?: Date;
  toDate?: Date;
  limit?: number;
  cursor?: LedgerCursor | null;
}

File: python/src/bursar/credits/types/ledger.py
"""Ledger types — mirrors JS SDK's ``credits/types/ledger.ts``."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Any, Literal, Self

from pydantic import BaseModel, model_validator


class LedgerEntry(BaseModel):
    entry_id: str
    account_id: str
    actor_user_id: str | None
    amount: Decimal
    entry_type: str
    operation: str
    reference_entry_id: str | None
    idempotency_key: str | None
    metadata: dict[str, Any] | None
    created_at: str


class LedgerCursor(BaseModel):
    created_at: str
    entry_id: str


class LedgerPage(BaseModel):
    items: list[LedgerEntry]
    next_cursor: LedgerCursor | None


class ListLedgerEntriesOptions(BaseModel):
    entry_types: list[str] | None = None
    from_date: datetime | None = None
    to_date: datetime | None = None
    limit: int | None = None
    cursor: LedgerCursor | None = None


class ListUsageEntriesOptions(BaseModel):
    from_date: datetime | None = None
    to_date: datetime | None = None
    limit: int | None = None
    cursor: LedgerCursor | None = None


class UsageCharge(BaseModel):
    usage_id: str
    account_id: str
    operation: str
    requested: Decimal
    charged: Decimal
    allowance_requested: Decimal
    allowance_covered: Decimal
    billing_disposition: Literal["billable", "record_only"]
    feature: str | None
    model: str | None
    region: str | None
    event_at: str
    idempotency_key: str
    metadata: dict[str, Any] | None
    created_at: str


class UsageChargeCursor(BaseModel):
    event_at: str
    usage_id: str


class UsageChargePage(BaseModel):
    items: list[UsageCharge]
    next_cursor: UsageChargeCursor | None


class UsageRecordResult(BaseModel):
    """Result of appending usage without creating another balance debit."""

    usage_id: str | None
    user_id: str
    requested: Decimal
    idempotent: bool
    error: str | None = None

    @model_validator(mode="after")
    def validate_outcome(self) -> Self:
        if self.error is None and self.usage_id is None:
            raise ValueError("successful usage records require a usage_id")
        if self.error is not None and (self.usage_id is not None or self.idempotent):
            raise ValueError("failed usage records cannot be committed or idempotent replays")
        return self


class ListUsageChargesOptions(BaseModel):
    from_date: datetime | None = None
    to_date: datetime | None = None
    limit: int | None = None
    cursor: UsageChargeCursor | None = None
    include_record_only: bool = True
